pkg/utils/requiredfieldsvalidator/validator.go

```diff
@@ -0,0 +1,139 @@
+package requiredfieldsvalidator
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+func ValidateRequiredFields(v any) error {
+	val := reflect.ValueOf(v)
+
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+
+	if val.Kind() != reflect.Struct {
+		return fmt.Errorf("expected struct type - got the %s", val.Kind())
+	}
+
+	err := validateValue(val)
+	if err != nil {
+		return fmt.Errorf("error have occured: %w", err)
+	}
+
+	return nil
+}
+
+func validateValue(val reflect.Value) error {
+	switch val.Kind() {
+	case reflect.Struct:
+		return validateStruct(val)
+
+	case reflect.Array, reflect.Slice:
+		return validateArrayOrSlice(val)
+
+	case reflect.Map:
+		return validateMap(val)
+
+	case reflect.Ptr:
+		return validatePointer(val)
+
+	case reflect.Bool:
+		return nil
+
+	default:
+		if val.IsZero() {
+			return fmt.Errorf("required field is empty")
+		}
+	}
+	return nil
+}
+
+func isStructPtr(fieldValue reflect.Value) bool {
+	return fieldValue.Kind() == reflect.Ptr && !fieldValue.IsNil() && fieldValue.Elem().Kind() == reflect.Struct
+}
+
+func tagContainsOmitempty(tag string) bool {
+	return strings.Contains(tag, models.OmitemptyTag)
+}
+
+func validateStruct(val reflect.Value) error {
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+		fieldValue := val.Field(i)
+
+		if (fieldValue.Kind() == reflect.Array || fieldValue.Kind() == reflect.Slice) && fieldValue.Len() > 0 {
+			err := validateValue(fieldValue)
+			if err != nil {
+				return fmt.Errorf("%s, %w", field.Name, err)
+			}
+		}
+
+		if tag, ok := field.Tag.Lookup(models.JsonTag); ok && tagContainsOmitempty(tag) {
+			if fieldValue.Kind() == reflect.Ptr || fieldValue.Kind() == reflect.Map {
+				if !fieldValue.IsZero() {
+					err := validateValue(fieldValue)
+					if err != nil {
+						return fmt.Errorf("%s, %w", field.Name, err)
+					}
+				}
+			}
+			continue
+		}
+
+		err := validateValue(fieldValue)
+		if err != nil {
+			return fmt.Errorf("%s, %w", field.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func validateArrayOrSlice(val reflect.Value) error {
+	if val.Len() < 1 {
+		return fmt.Errorf("len is 0")
+	}
+	for i := 0; i < val.Len(); i++ {
+		fieldValue := val.Index(i)
+		if fieldValue.Kind() == reflect.Struct ||
+			isStructPtr(fieldValue) {
```

Comment:
I see you reuse the following code multiple times. What do you think to wrap this logic inside the func?

```go
func structValOrPtr(val reflect.Value) bool {
  return val.Kind() == reflect.Struct || isStructPtr(val)
}
```

Reply:
agree, refactored